pkg/api/v1/pod/util.go

```diff
@@ -50,6 +50,59 @@ func FindPort(pod *v1.Pod, svcPort *v1.ServicePort) (int, error) {
 	return 0, fmt.Errorf("no suitable port for manifest: %s", pod.UID)
 }
 
+// ContainerType specifies the type of container
+type ContainerType int
+
+const (
+	// DontCheckFeatureGate skips checking feature gate
+	DontCheckFeatureGate = 1 << iota
+	// InitContainer is for init container
+	InitContainer
+	// Container is for container
+	Container
+	// EphemeralContainer is for ephemeral container which checks feature gate
+	EphemeralContainer
+	// AllContainers cover all containers
+	AllContainers = InitContainer | Container | EphemeralContainer
+	// NonEphemeralContainers cover all containers except ephemeral container
+	NonEphemeralContainers = InitContainer | Container
+)
+
+// ContainerVisitorV2 is called with container spec of given container type, and returns true
+// if visiting should continue.
+type ContainerVisitorV2 func(container *v1.Container, containerType ContainerType) (shouldContinue bool)
```

Comment:
if we're making a v2 visitor signature, passing in the index as well makes this useful for validation methods to construct fieldpaths, etc